### ISSUE TITLE
Encounter a infinite loop seems related to :not(:hover)

### ISSUE THREAD

**Issue by lackneets** [saulhardman/postcss-hover-media-feature#20]
## What happened?

I found my webpack-dev-server startup hangs on with 100% cpu usage forever. To find out what happened, I modify the source code with console.log to see what's going on. As a result, It stuck on this rule, keeping looping

![image](https://user-images.githubusercontent.com/3694058/149060012-f86b41c7-7122-412e-9ae7-ef3fcb01f3c3.png)

![image](https://user-images.githubusercontent.com/3694058/149061900-19c5c543-94d5-4e80-bb31-9053d0364c2b.png)

## To reproduce this problem

![image](https://user-images.githubusercontent.com/3694058/149060091-aea93ed2-07f0-47c8-8545-6cdc22292612.png)

### source (stylus)

```stylus
.list
  margin 0
  &__item, &__action
    padding .25rem 1rem
    margin-bottom 1rem
    border-radius 5px
    font-size 17px
  &__item
    box-shadow 0px 1px 1px 0 rgba(170, 170, 170, .2)
    border solid 1px $gray3
    background-color #ffffff
    padding 1rem
    .column
      line-height 2
    .column:first-child
      display flex !important
      line-height 2
      align-items flex-start
      > *
        vertical-align middle
      > .ui-checkbox
        align-self flex-start
        min-height 34px
        display inline-flex
        align-items center

    &:not(:hover):not(.is-editing) .show-on-hover
      visibility hidden
    &:hover
      background-color #f0f8ff
  &__action
    color $gray2
    cursor pointer
    &:hover
      background $gray3

.nav-top
  padding 21px

div.table-list
  display flex
  position relative
  flex-direction column
  justify-content flex-start
  margin 0
  .tr
    padding-left 21px
    padding-right 21px
    &.checked
      background-color #FEF9F4
  .th, .td
    padding .25rem .5rem
  .tbody
    height 100%
  /deep/ .vue-recycle-scroller__item-wrapper
    height 100%
  .tr
    display flex
    width 100%
    min-height 45px
    align-items center
    border-bottom 1px solid $gray4
  > .thead, > .tr
    margin-right 10px
    width auto
  .th
    color $gray2
    font-size 12px
  .center
    text-align center
  .right
    text-align right

.dot-status
  width 5px
  height 5px
  display inline-block
  vertical-align middle
  border-radius 50%
  background-color $gray3
  &--active
    background-color #ff895a

.auto-scroll-y
  overflow-y scroll !important
  overflow-x hidden
  max-height 100%
```

**Comment by lackneets** [comment]
## My attempt

After I realized that the selector inside :not(), will be split into child selector, that won't be walked through. 

Finally I replace it with a recursive method:

![image](https://user-images.githubusercontent.com/3694058/149063511-255f730f-7327-4496-85fb-4ec8411b9828.png)


## Tada!

![image](https://user-images.githubusercontent.com/3694058/149063699-4fa4aeca-5d82-4ccf-89c0-aa963a1b3294.png)

![image](https://user-images.githubusercontent.com/3694058/149063343-b6e0bac8-c077-489c-8942-3ed602814399.png)

**Comment by dustinwilson** [comment]
I also ran into this issue today wondering why there was an infinite loop, and sure enough it was `:not(:hover)` just like you discovered.

**Comment by dustinwilson** [comment]
Although, I think it should instead ignore `:not(:hover)` because in a user agent without hover support the selector would (and should) match.

**Comment by saulhardman** [comment]
Hey @lackneets and @dustinwilson, thanks for the detailed bug report and confirmation. I'm aiming to take a look at this now and get a release out in the next few hours. As @dustinwilson suggested; I'll change the selector processor to omit selectors where `:hover` is within a `:not` or some variation of that logic.

I'll let you know when a patch is out there ready to test 👍

**Comment by saulhardman** [comment]
Hello again – I've just released a patch which hopefully resolves the issue. I had to do a manual release as it looks like `travis-ci.org` is no longer operational. I guess that means I'll have to spend some more time migrating to GitHub Actions sometime soon.

Regarding the source of this problem: the way that selectors are identified, parsed, and rewritten definitely needs a rethink at some point, but sadly I don't have the time to dive into it right now.

Thanks again for the bug report, confirmation, and suggestion of solutions 🙏
